File: rust/src/models/product_ref.rs
// this file is @generated
use serde::{Deserialize, Serialize};

use super::{existing_product_ref::ExistingProductRef, new_product_ref::NewProductRef};

#[derive(Clone, Debug, PartialEq, Deserialize, Serialize)]
#[serde(tag = "type")]
pub enum ProductRef {
    #[serde(rename = "EXISTING")]
    Existing(ExistingProductRef),
    #[serde(rename = "NEW")]
    New(NewProductRef),
}

impl Default for ProductRef {
    fn default() -> Self {
        Self::Existing(ExistingProductRef::default())
    }
}

File: typescript/src/models/productRef.ts
// this file is @generated
import {
  type ExistingProductRef,
  ExistingProductRefSerializer,
} from "./existingProductRef";
import { type NewProductRef, NewProductRefSerializer } from "./newProductRef";

export interface ProductRefExisting extends ExistingProductRef {
  type: "EXISTING";
}
export interface ProductRefNew extends NewProductRef {
  type: "NEW";
}

export type ProductRef = ProductRefExisting | ProductRefNew;

export const ProductRefSerializer = {
  _fromJsonObject(object: any): ProductRef {
    const type = object["type"];

    switch (type) {
      case "EXISTING":
        return {
          ...ExistingProductRefSerializer._fromJsonObject(object),
          type: "EXISTING",
        };
      case "NEW":
        return {
          ...NewProductRefSerializer._fromJsonObject(object),
          type: "NEW",
        };
      default:
        throw new Error(`Unexpected type for ProductRef: ${type}`);
    }
  },

  _toJsonObject(self: ProductRef): any {
    switch (self.type) {
      case "EXISTING":
        return {
          ...ExistingProductRefSerializer._toJsonObject(self),
          type: "EXISTING",
        };
      case "NEW":
        return {
          ...NewProductRefSerializer._toJsonObject(self),
          type: "NEW",
        };
      default:
        throw new Error(`Unexpected type for ProductRef`);
    }
  },
};
